usr/lib/soft_stdll/soft_specific.c

```diff
@@ -3172,19 +3163,86 @@ CK_RV token_specific_sha_init(STDLL_TokData_t *tokdata, DIGEST_CONTEXT *ctx,
         break;
     }
 
+    return md;
+}
+
+static EVP_MD_CTX *md_ctx_from_context(DIGEST_CONTEXT *ctx)
+{
+    const EVP_MD *md;
+    EVP_MD_CTX *md_ctx;
+
+#if OPENSSL_VERSION_NUMBER < 0x10101000L
+    md_ctx = EVP_MD_CTX_create();
+#else
+    md_ctx = EVP_MD_CTX_new();
+#endif
+    if (md_ctx == NULL)
+        return NULL;
+
+    md = md_from_mech(&ctx->mech);
     if (md == NULL ||
-        !EVP_DigestInit_ex((EVP_MD_CTX *)ctx->context, md, NULL)) {
+        !EVP_DigestInit_ex(md_ctx, md, NULL)) {
+        TRACE_ERROR("md_from_mech or EVP_DigestInit_ex failed\n");
 #if OPENSSL_VERSION_NUMBER < 0x10101000L
-        EVP_MD_CTX_destroy((EVP_MD_CTX *)ctx->context);
+        EVP_MD_CTX_destroy(md_ctx);
 #else
-        EVP_MD_CTX_free((EVP_MD_CTX *)ctx->context);
+        EVP_MD_CTX_free(md_ctx);
 #endif
-        ctx->context = NULL;
-        ctx->context_len = 0;
+        return NULL;
+    }
 
-        return CKR_FUNCTION_FAILED;
+    if (ctx->context_len == 0) {
+        ctx->context_len = EVP_MD_meth_get_app_datasize(EVP_MD_CTX_md(md_ctx));
+        ctx->context = malloc(ctx->context_len);
+        if (ctx->context == NULL) {
+            TRACE_ERROR("malloc failed\n");
+    #if OPENSSL_VERSION_NUMBER < 0x10101000L
+            EVP_MD_CTX_destroy(md_ctx);
+    #else
+            EVP_MD_CTX_free(md_ctx);
+    #endif
+            ctx->context = NULL;
```

Comment:
Assignment useless in this branch since `ctx->context == NULL` is the only condition leading here.

Reply:
fixed